wire-ios-request-strategy/Sources/Message Sending/MessageSenderTests.swift

```diff
@@ -117,8 +121,88 @@ final class MessageSenderTests: MessagingTestBase {
+
+        let (_, messageSender) = Arrangement(coreDataStack: coreDataStack)
+            .withQuickSyncObserverCompleting()
+            .withApiVersionResolving(to: .v0)
+            .withBroadcastProteusMessage(returning: .success((messageSendingStatus, response)))
+            .arrange()
+
+        // when
+        try await messageSender.broadcastMessage(message: message)
+
+        // then test completes
+    }
+
+    func testThatWhenBroadcastingProteusMessageFailsDueToMissingClients_thenEstablishSessionsAndTryAgain() async throws {
+        // given
+        let response = ZMTransportResponse(payload: nil, httpStatus: 412, transportSessionError: nil, apiVersion: 0)
+        let message = GenericMessageEntity(
+            message: GenericMessage(content: Text(content: "Hello World")),
+            context: syncMOC,
+            conversation: groupConversation,
+            completionHandler: nil)
+
+        let (arrangement, messageSender) = Arrangement(coreDataStack: coreDataStack)
+            .withQuickSyncObserverCompleting()
+            .withApiVersionResolving(to: .v0)
+            .withBroadcastProteusMessageFailing(with: NetworkError.missingClients(
+                Arrangement.Scaffolding.messageSendingStatusMissingClients,
+                response)
+            )
+            .withEstablishSessions(returning: .success(Void()))
+            .arrange()
+
+        // when
+        try await messageSender.broadcastMessage(message: message)
+
+        // then
+        XCTAssertEqual(Set(arrayLiteral: Arrangement.Scaffolding.clientID), arrangement.sessionEstablisher.establishSessionWithApiVersion_Invocations[0].clients)
+    }
+
+    func testThatWhenBroadcastingMessageProteusFailsWithTemporaryError_thenTryAgain() async throws {
+        // given
+        let response = ZMTransportResponse(payload: nil, httpStatus: 408, transportSessionError: nil, apiVersion: 0)
+        let message = GenericMessageEntity(
             message: GenericMessage(content: Text(content: "Hello World")),
+            context: syncMOC,
+            conversation: groupConversation,
+            completionHandler: nil)
+
+        let (arrangement, messageSender) = Arrangement(coreDataStack: coreDataStack)
+            .withQuickSyncObserverCompleting()
+            .withApiVersionResolving(to: .v0)
+            .withBroadcastProteusMessageFailing(with: NetworkError.errorDecodingResponse(response))
+            .withEstablishSessions(returning: .success(Void()))
+            .arrange()
+
+        // when
+        try await messageSender.broadcastMessage(message: message)
+
+        // then
+        XCTAssertEqual(2, arrangement.messageApi.broadcastProteusMessageMessage_Invocations.count)
+    }
+
+    func testThatWhenSendingProteusMessageSucceeds_thenCompleteWithoutErrors() async throws {
+        // given
+        let response = ZMTransportResponse(payload: nil, httpStatus: 200, transportSessionError: nil, apiVersion: 0)
```

Comment:
does it make sense to test for other apiVersion?

Reply:
No, it's the same logic